server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -282,6 +282,10 @@ static boolean execute(SearchContext searchContext,
             } finally {
                 searchContext.clearReleasables(SearchContext.Lifetime.COLLECTION);
             }
+            if (searchContext.terminateAfter() != SearchContext.DEFAULT_TERMINATE_AFTER
+                    && queryResult.terminatedEarly() == null) {
```

Comment:
I find this logic a bit confusing, would it be possible to have only true or false in QuerySearchResult? What is the difference between null and false again?

Reply:
In 6x we use `true` if the search was early terminated due to `terminate_after` and `null` in all other cases. However in 5x we set it to `false` if `terminate_after` is used but the number of documents that match the query is smaller than the terminate_after value. We broke this behavior in 6x which returns `null` in this case. I think it's less confusing to always set the early terminate flag if `terminate_after` is set in the query, this way users can rely on the fact that it will not be null in the response.

Reply:
I agree that it is less confusing this way, and I was wondering if it could be made even less confusing by removing the null invariant from QuerySearchResult. I take it that it's more complicated than it seems though.